### PR TITLE
[Feature]Support ignore nulls for lead/lag window function

### LINKED ISSUE #16350: Support "IGNORE NULL" option for LEAD and LAG
## Feature request

**Is your feature request related to a problem? Please describe.**
Currently LEAD and LAG take into account NULLs whereas the current semantic that we need to emulate for backwards compatibility is IGNORE NULL.

**Describe the solution you'd like**
Would like an IGNORE NULL option as described here:
https://docs.snowflake.com/en/sql-reference/functions/lag.html

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by before-Sunrise**
Signed-off-by: before-Sunrise <unclejyj@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #16350

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exprs/agg/window.h [hunk 1]
```diff
@@ -400,7 +399,7 @@ struct LeadLagState {
     bool defualt_is_null = false;
 };
 
-template <LogicalType PT, typename T = RunTimeCppType<PT>>
+template <LogicalType PT, bool ignoreNulls, bool use_lag, typename T = RunTimeCppType<PT>>
```
Comment (mofeiatwork) [inline be/src/exprs/agg/window.h:402]: unify the naming style?

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by before-Sunrise** [comment]
run starrocks_admit_test

File: be/src/exprs/agg/factory/aggregate_factory.hpp [hunk 2]
```diff
@@ -182,9 +182,9 @@ class AggregateFactory {
         return std::make_shared<LastValueWindowFunction<PT, ignoreNulls>>();
     }
 
-    template <LogicalType PT>
+    template <LogicalType PT, bool ignoreNulls, bool useLag>
```
Comment (liuyehcf) [inline be/src/exprs/agg/factory/aggregate_factory.hpp:185]: useLag -> isLag

File: be/src/exprs/agg/window.h [hunk 3]
```diff
@@ -432,16 +431,33 @@ class LeadLagWindowFunction final : public ValueWindowFunction<PT, LeadLagState<
...
+            const InputColumnType* column = down_cast<const InputColumnType*>(data_column);
+            AggDataTypeTraits<PT>::assign_value(this->data(state).value,
+                                                AggDataTypeTraits<PT>::get_row_ref(*column, frame_end - 1));
+        } else {
+            if (!ignoreNulls) {
+                this->data(state).is_null = true;
+                return;
+            }
+            // for lead/lag, [peer_group_start, peer_group_end] equals to [partition_start, partition_end]
+            // when lead/lag called, the whole partitoin's data already been here, so we can just check to the begining or the end
```
Comment (liuyehcf) [inline be/src/exprs/agg/window.h:446]: already been here -> has already been here
check to the begining -> check the begining

File: be/src/exprs/agg/window.h [hunk 4]
```diff
@@ -432,16 +431,33 @@ class LeadLagWindowFunction final : public ValueWindowFunction<PT, LeadLagState<
...
+            size_t value_index = useLag ? ColumnHelper::last_nonnull(columns[0], peer_group_start, frame_end - 1)
+                                        : ColumnHelper::find_nonnull(columns[0], frame_end, peer_group_end);
```
Comment (liuyehcf) [inline be/src/exprs/agg/window.h:448]: why `ColumnHelper::find_nonnull(columns[0], frame_end, peer_group_end);` ?

I suppose `frame_end >= peer_group_end` is always positive?

Comment (before-Sunrise) [reply be/src/exprs/agg/window.h:448]: for lead/lag, peer_group_end is equal to partition_end, so peer_group_end >= frame_end is always positive. I just use peer_group_end as  partition_end

Comment (liuyehcf) [reply be/src/exprs/agg/window.h:448]: Ok, I misunderstood.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by murphyatwork — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by before-Sunrise** [comment]
@mergify backport branch-2.5

**Pull request merged**